File: frontend/src/api/featureFlags.ts
/**
 * Feature-flag API client.
 *
 * Mirrors `backend/app/api/v1/feature_flags.py`. Public endpoint — no token
 * required so the login page can branch before authentication.
 */
import api from './client'

export interface FeatureFlags {
  osa_flow_ui_enabled: boolean
  /**
   * Gates the Kali coexistence tool surface (kali_gobuster, kali_sqlmap,
   * kali_nikto). Default false. When false the backend hides kali_* from
   * /agents/catalog and refuses get_adapter; the UI uses this to render
   * a "Kali disabled" notice if a workflow still references those slugs.
   */
  osa_kali_backend_enabled: boolean
  /** Gates multi-LLM credential management UI and provider routing. Default false. */
  osa_multi_provider_llm?: boolean
  /** Gates the Coordinator split-pane UI (understanding + plan-of-work + conversation). Default false. */
  osa_coordinator_enabled?: boolean
  /** Gates AgentFamilyTree UI and family-spawning API surface. Default false. */
  osa_xbow_families_enabled?: boolean
}

export async function getFeatureFlags(): Promise<FeatureFlags> {
  const { data } = await api.get<FeatureFlags>('/config/feature-flags')
  return data
}

File: backend/app/api/v1/feature_flags.py
"""Feature flag API (v4.0 P3-main).

Single endpoint that surfaces the v4.0 flag set to the frontend so the React
app can branch between the legacy 12-page layout and the new /flow/:id shell.
The flag itself lives in `settings.osa_flow_ui_enabled` (default False); the
endpoint reads from that and never sets it — flag flips happen out-of-band
(env var or settings override).

ADR-004 scope: global flag (no per-team / per-user override in v4.0).
"""
from __future__ import annotations

from fastapi import APIRouter

from app.core.config import settings

router = APIRouter()


@router.get("/config/feature-flags")
async def get_feature_flags() -> dict[str, bool]:
    """Return the public feature-flag set. No auth required so the login
    page (and pre-login screens) can branch before authentication.

    - osa_flow_ui_enabled: switches the /flow/:id 2-pane shell on/off
    - osa_kali_backend_enabled: gates the kali_* tool surface; UI uses it
      to decide whether to render a "Kali disabled" notice and to hide
      kali_* options that the backend would also hide
    """
    return {
        "osa_flow_ui_enabled": settings.osa_flow_ui_enabled,
        "osa_kali_backend_enabled": settings.osa_kali_backend_enabled,
    }
